Congrats: Refactor congrats page for wpcom and p2+ plans

wongasy opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/5125

## Proposed Changes

This PR refactors the Plan-only Congrats Page to adapt the new approach for unifying the congratulation pages using the `<ThankYouV2>` component. See pb5gDS-38y-p2#comment-4475 for details.

The Plan-only Congrats Page is the last redesign page that needs to be refactored. So in addition to the refactoring work, this PR also removes redesign files / code that are no longer needed.

Note that, the placeholder skeleton is still rendered using the legacy component. Once we have a good number of pages ported over to the redesign, we should update the following logic to render a placeholder using `<ThankYouV2>`. https://github.com/Automattic/wp-calypso/blob/761d303334cd33e64dfadc2e1914919526c76cce/client/my-sites/checkout/checkout-thank-you/index.tsx#L876-L901

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### WPCOM plan

* If you don't have a plan previously purchased, purchase one.
* If you have a plan handy, get the respective receipt ID using the following steps:
  * Go to Upgrades > Purchases > Billing History
  * Look for the plan purchase and click on "View receipt" and you should be able to find the receipt ID
* Visit `/checkout/thank-you/:site/:receipt_id` and assert that you see the Plan-only Congrats Page. Compare the same page in production and assert that you see the same page content (note that styling will be slightly different, since we are now using `<ThankYouV2>` for a unified layout).
* If the plan was purchased for a site that is 6+ months old, then you should see two detail blocks in the footer, including a blurb for site refresh.

<img width="857" alt="Screen Shot 2024-01-15 at 12 14 26 PM" src="https://github.com/Automattic/wp-calypso/assets/104910361/c84d10a7-d602-4303-bffd-c5ce264e7fe2">

* If the plan was purchased for a site that is less than 6 months old, then you should only see one detail block in the footer (the site refresh blurb should not be shown).

<img width="860" alt="Screen Shot 2024-01-15 at 12 13 55 PM" src="https://github.com/Automattic/wp-calypso/assets/104910361/c4566751-fd1c-4883-ad6c-212230b968ed">

* You may adjust the site age by sandboxing `public-api.wordpress.com` and hardcoding an appropriate date using a similar diff 332d9-pb

### P2+ plan

* If you don't have a P2+ plan previously purchased, purchase one. To do so, go to `https://wordpress.com/start/p2`, you should be able to set up a P2 for free. Once you have a P2, upgrade it to P2 plus.
* If you have a P2+ plan handy, get the respective receipt ID from your billing history.
* Visit `/checkout/thank-you/:site/:receipt_id` and assert that you see the Plan-only Congrats Page, where the first detail block should say "Go further, together", regardless of site age. 

<img width="864" alt="Screen Shot 2024-01-15 at 12 19 44 PM" src="https://github.com/Automattic/wp-calypso/assets/104910361/ce38621f-6282-43e5-92df-e594de5ef400">

Note that, while production redesignV2 code includes [P2+ specific code](https://github.com/Automattic/wp-calypso/blob/d5b4472c3505199dcb06df5a5ac9df058e68661b/client/my-sites/checkout/checkout-thank-you/redesign-v2/sections/footer/P2PlusPlanFooter.tsx#L14-L22), the Congrats page for P2+ plan appears to be on legacy design in production. @candy02058912 Please let us know if we've missed anything? Have we ever applied the redesign to the Congrats Page for P2+ plans in production? 

<img width="815" alt="Screen Shot 2024-01-15 at 12 21 17 PM" src="https://github.com/Automattic/wp-calypso/assets/104910361/2bcc5d0e-13f7-4bad-bdb6-da5f4c137568">

wongasy committed:
Refactored domain only congrats page

wongasy committed:
Fixed type errors

wongasy committed:
Type fixes

wongasy committed:
Updated back button label

wongasy committed:
Refactor congrats page for wpcom plans

wongasy committed:
Add footer details keys

wongasy committed:
Removed unused footer

wongasy committed:
Updated back button label as per mock

wongasy committed:
Tweaked style and clean up

wongasy committed:
Removed duplciate import

wongasy committed:
removed component that was re-added during rebase

wongasy committed:
Updated unit tests

wongasy committed:
Support p2+ plans in refactored plan-only congrats page

wongasy commented on client/components/thank-you-v2/footer/index.tsx:
```diff
@@ -54,7 +55,11 @@ type ThankYouFooterProps = {
-		<div className="thank-you__footer">
+		<div
+			className={ classNames( 'thank-you__footer', {
+				'has-multiple-details': details && details.length > 1,
```
This allows us to port over this feature https://github.com/Automattic/wp-calypso/pull/85478 for rendering footer details in full width when there's only one single blurb to be rendered.

wongasy committed:
renamed prop for clarity

wongasy committed:
Removed redesign v2 files that are no longer needed

wongasy committed:
Clean up redesign v2 styles that's no longer needed

wongasy committed:
Removed old file that was added by accident during rebase

wongasy committed:
Updated unit test to cover p2 plus plans

wongasy committed:
Fixed unit test

wongasy committed:
Fixed type errors

wongasy committed:
Fixed type error

wongasy committed:
Set is loading state properly

wongasy commented on client/my-sites/checkout/checkout-thank-you/redesign-v2/pages/plan-only.tsx:
```diff
@@ -0,0 +1,93 @@
+				)
+			) }
+			products={
+				<ThankYouPlanProduct purchase={ primaryPurchase } siteSlug={ siteSlug } siteId={ siteId } />
```
Noting that I've pivoted from the previous decision made (https://github.com/Automattic/wp-calypso/pull/86201#discussion_r1450789600) to always pass an array. Reason being, in the context of site plans, users would only purchase one single plan at a time, supplying the single purchase as an array incorrectly suggest that we may supply multiple plan products in the future. 

The Domain Transfer to User Congrats Page has also been updated accordingly to pass the single product without array wrappers.

wongasy commented on client/my-sites/checkout/checkout-thank-you/domain-transfer-to-any-user/index.tsx:
```diff
@@ -34,18 +34,16 @@ const DomainTransferToAnyUser: React.FC< DomainTransferToAnyUserContainerProps >
 					'Domain transfers can take a few minutes, we’ll email you once it’s set up.'
 				) }
-				products={ products }
+				products={ <ThankYouDomainProduct domainName={ domain } key="transferred-domain" /> }
```
See https://github.com/Automattic/wp-calypso/pull/86271/files#r1452631217

nightnei replied:
```suggestion
				products={ <ThankYouDomainProduct domainName={ domain } /> }
```

`key` is unnecessary for arrays

wongasy commented on client/my-sites/checkout/checkout-thank-you/header.jsx:
```diff
@@ -511,21 +508,45 @@ export class CheckoutThankYouHeader extends PureComponent {
-		return getHeading( {
-			...this.props,
-			isSearch: this.isSearch(),
-		} );
```
This was previously refactored to a separate file as part of the redesign: https://github.com/Automattic/wp-calypso/pull/78931/files#r1254074573 Now that we're pivoting to a different approach pb5gDS-38y-p2#comment-4475, I've moved that logic back to the legacy header, so it's easier to identify what needs to be cleaned up / removed in the future.

wongasy commented on client/my-sites/checkout/checkout-thank-you/header.jsx:
```diff
@@ -59,7 +57,6 @@ export class CheckoutThankYouHeader extends PureComponent {
 		translate: PropTypes.func.isRequired,
 		_n: PropTypes.func.isRequired,
 		upgradeIntent: PropTypes.string,
-		isRedesignV2: PropTypes.bool,
```
We no longer need the `isRedesignV2` check for rendering `<CheckoutThankYouHeader>`. All redesigned congrats pages should not be using `<CheckoutThankYouHeader>` at all.

wongasy commented on client/my-sites/checkout/checkout-thank-you/index.tsx:
```diff
@@ -712,24 +712,8 @@ export class CheckoutThankYou extends Component<
-			>
+			<Main className="checkout-thank-you">
 				<PageViewTracker { ...this.getAnalyticsProperties() } title="Checkout Thank You" />
-				{ this.isDataLoaded() && isRedesignV2( this.props ) && (
```
We no longer need the redesign check here. All redesigned pages should render using the logic defined by these lines: https://github.com/Automattic/wp-calypso/pull/86271/files#diff-daac408bb350333e9fb8ca7c03233e75f03656e02a45c152e574227e6d253ee7R569-R610

wongasy commented on client/my-sites/checkout/checkout-thank-you/index.tsx:
```diff
@@ -913,10 +897,9 @@ export class CheckoutThankYou extends Component<
-					isRedesignV2={ isRedesignV2( this.props ) }
 					currency={ receipt.data?.currency }
 				>
-					{ ! isRedesignV2( this.props ) && ! isSimplified && primaryPurchase && (
```
Similarly, we no longer need the redesign check here. All redesigned pages should render using the logic defined by these lines: https://github.com/Automattic/wp-calypso/pull/86271/files#diff-daac408bb350333e9fb8ca7c03233e75f03656e02a45c152e574227e6d253ee7R569-R610

wongasy commented on client/my-sites/checkout/checkout-thank-you/index.tsx:
```diff
@@ -927,14 +910,10 @@ export class CheckoutThankYou extends Component<
 
 					{ ! isSimplified && component && (
 						<div className="checkout-thank-you__purchase-details-list">
-							{ isRedesignV2( this.props ) ? (
```
Similarly, see https://github.com/Automattic/wp-calypso/pull/86271/files#r1452638442

fredrikekelund commented on client/my-sites/checkout/checkout-thank-you/redesign-v2/pages/plan-only.tsx:
```diff
@@ -0,0 +1,93 @@
+	primaryPurchase: ReceiptPurchase;
+}
+
+export default function PlanOnlyThankYou( { primaryPurchase }: PlanOnlyThankYouProps ) {
+	const isMonthsOld = ( rawDate: string, months: number ) => {
+		if ( ! rawDate || ! months ) {
+			return false;
+		}
+
+		const parsedDate = moment( rawDate );
+		return moment().diff( parsedDate, 'months' ) > months;
+	};
```
```suggestion
const isMonthsOld = ( rawDate: string, months: number ) => {
	if ( ! rawDate || ! months ) {
		return false;
	}

	const parsedDate = moment( rawDate );
	return moment().diff( parsedDate, 'months' ) > months;
};

export default function PlanOnlyThankYou( { primaryPurchase }: PlanOnlyThankYouProps ) {
```
No need to define `isMonthsOld` within the component.

nightnei replied:
Also take a look please at my comment below with explanation regardinng optional parameters

fredrikekelund commented on client/components/thank-you-v2/index.tsx:
```diff
@@ -10,7 +10,7 @@ interface ThankYouV2Props {
 	subtitle: TranslateResult;
 	headerButtons?: React.ReactNode;
-	products?: React.ReactElement[];
+	products?: React.ReactElement | React.ReactElement[];
```
```suggestion
	products?: React.ReactNode;
```
`ReactNode` allows (among other things) both single React components and arrays of components. Maybe it makes sense to use it here?

nightnei replied:
Discussion behind it: https://github.com/Automattic/wp-calypso/pull/86201#discussion_r1450740520

Also React`ReactNode` allows `string`s and other types, so it literally looks like `any`. I would be more specific with `ReactElement `.

fredrikekelund commented on client/components/thank-you-v2/style.scss:
```diff
@@ -14,4 +14,9 @@
 		width: 100%;
 	}
+
+	.components-button {
+		border-radius: 4px;
+		padding: 10px 24px;
+	}
```
Nice 👍 I overrode this in #86409 for the header (see https://github.com/Automattic/wp-calypso/pull/86409#discussion_r1452473414). I'll update that PR to account for this incoming change

fredrikekelund commented on client/my-sites/checkout/checkout-thank-you/redesign-v2/style.scss:
WDYT about potentially moving the contents of this file into `client/my-sites/checkout/checkout-thank-you/style.scss`? It seems like a potentially more manageable strategy to me. It wasn't even clear to me initially which component is importing this file now.

wongasy replied:
Yeah, I've thought about that as well, but ended up opting to keep the `checkout/checkout-thank-you/redesign-v2/style.scss` just to avoid the need to distinguish between old/new design when we do get to clean up `checkout/checkout-thank-you/style.scss`. However, given the redesign css definition is minimal, it's probably fine to include it in the original `style.scss` to avoid confusion around where the import takes place https://github.com/Automattic/wp-calypso/pull/86271/commits/d9746fa59fbad9f33a84578e4efcc57538335009

fredrikekelund reviewed (approved):
Looks great and tests well 👍 I left a few comments, but nothing blocking.

nightnei commented on client/my-sites/checkout/checkout-thank-you/header.jsx:
```diff
@@ -511,21 +508,45 @@ export class CheckoutThankYouHeader extends PureComponent {
-		} );
+		const { isDataLoaded, hasFailedPurchases, primaryPurchase, translate } = this.props;
+
+		if ( ! isDataLoaded ) {
```
Not critical, but I would use `if ( isDataLoading ) {`

Since now it looks a bit confusing, the condition below is literally - `isDataLoaded && hasFailedPurchases`, but actually it's not loaded, it's failed . Hope you get what I mean :) 
But not critical, up to you.

UPDATE: Oh, I see the method... ignore this my comment :)

wongasy replied:
Also noting that, ultimately, we're looking to clean up (or remove) `header.jsx` at some point. Currently, it's only being used for legacy congrats pages that aren't redesigned.

nightnei commented on client/my-sites/checkout/checkout-thank-you/redesign-v2/pages/plan-only.tsx:
```diff
@@ -0,0 +1,93 @@
+				recordTracksEvent( 'calypso_plan_thank_you_add_members_click' );
+			},
+		} );
+	} else if ( isMonthsOld( siteCreatedTimeStamp ?? '', 6 ) ) {
```
```suggestion
	} else if ( isMonthsOld( siteCreatedTimeStamp, 6 ) ) {
```

in the isMonthsOld we have condition `if ( ! rawDate || ! months ) {`, so let's go the next way:
1) `const isMonthsOld = ( rawDate?: string, months?: number ) => {` // allow undefined
2) `} else if ( isMonthsOld( siteCreatedTimeStamp, 6 ) ) {` provide undefined instead of forcing empty string

wongasy replied:
Good point! I also don't see why we wouldn't supply `months`, so I've updated the function as follows:

```
const isMonthsOld = ( rawDate?: string, months: number ) => {
	if ( ! rawDate ) {
		return false;
	}

	const parsedDate = moment( rawDate );
	return moment().diff( parsedDate, 'months' ) > months;
}
```

The optional parameter `rawDate` is really just meant to account for the case where `siteCreatedTimeStamp` is unexpectedly missing.

nightnei commented on client/my-sites/checkout/checkout-thank-you/redesign-v2/products/plan-product.tsx:
```diff
@@ -0,0 +1,71 @@
+
+	const [ expirationDate, setExpirationDate ] = useState( '' );
+
+	useEffect( () => {
+		if ( ! isLoadingPurchases && productPurchase ) {
+			setExpirationDate(
+				translate( 'Expires on %s', {
+					args: moment( productPurchase.expiryDate ).format( 'LL' ),
+				} ).toString()
+			);
+		}
+	}, [ isLoadingPurchases, productPurchase ] );
```
This `useEffect` andf storing translation in state looks like overkill. And I can't get the reason to do it, instead of simply using something like

```
const expirationDate = ! isLoadingPurchases && productPurchase ?
    ? translate( 'Expires on %s', { args: moment( productPurchase.expiryDate ).format( 'LL' ) } )
    : '';
```

Or did I miss something?

nightnei reviewed (changes requested):
In general LGTM, but I have a few not critical comments

wongasy committed:
Addressed feedback

wongasy committed:
Prioritize mandatory parameter

wongasy committed:
Removed standalone redesign scss file

wongasy commented:
@fredrikekelund @nightnei I've addressed the feedback! There's one thing that I wasn't able to figure out and wonder if you'd be able to help. The confetti animation animates twice when I load up the Plan Congrats Page, however, that isn't the case for the Domain Congrats Page. Do you have any idea why that is and how to go about fixing it?

fredrikekelund commented:
> The confetti animation animates twice

This happens occasionally for me, but not consistently. Nothing sticks out to me in the code, so I would probably just file it under "occasional glitch in dev environment".

nightnei commented:
>The confetti animation animates twice

@wongasy It was interesting thing, so I investigated and figured out the issue. Results:
1) In a nutshel - we are good to deploy the changes from this PR, since the issue was before. Totally nothing changes in the behaviour.
2) The issue exists not only in development, but also in production.
3) It's not a simple bug to explain with the help of a few sentences, so I will record the video with details, I think it will be interesting and maybe even useful :)

wongasy commented:
Thanks @fredrikekelund and @nightnei for looking into this! I'll go ahead and deploy after rebasing.
